graphql_compiler/schema_generation/sqlalchemy/edge_descriptors.py

```diff
@@ -1,64 +1,120 @@
+    # and to_column is on the destination table of the join.
+    column_pairs: AbstractSet[Tuple[str, str]]
+
+    def __post_init__(self) -> None:
+        """Validate fields."""
+        if not self.column_pairs:
+            raise AssertionError("The column_pairs field is expected to be non-empty.")
+
+
+EdgeDescriptor = Union[DirectEdgeDescriptor, CompositeEdgeDescriptor]
+
+
 def get_join_descriptors_from_edge_descriptors(
-    direct_edges: Dict[str, DirectEdgeDescriptor]
-) -> Dict[str, Dict[str, DirectJoinDescriptor]]:
+    direct_edges: Dict[str, EdgeDescriptor]
+) -> Dict[str, Dict[str, JoinDescriptor]]:
     """Return the SQL edges in a format more suited to resolving vertex fields."""
-    join_descriptors: Dict[str, Dict[str, DirectJoinDescriptor]] = {}
-    for edge_name, direct_edge_descriptor in direct_edges.items():
-        from_column = direct_edge_descriptor.from_column
-        to_column = direct_edge_descriptor.to_column
-        join_descriptors.setdefault(direct_edge_descriptor.from_vertex, {})
-        join_descriptors.setdefault(direct_edge_descriptor.to_vertex, {})
+    join_descriptors: Dict[str, Dict[str, JoinDescriptor]] = {}
+    for edge_name, edge_descriptor in direct_edges.items():
+        join_descriptors.setdefault(edge_descriptor.from_vertex, {})
+        join_descriptors.setdefault(edge_descriptor.to_vertex, {})
         out_edge_name = OUTBOUND_EDGE_FIELD_PREFIX + edge_name
         in_edge_name = INBOUND_EDGE_FIELD_PREFIX + edge_name
-        join_descriptors[direct_edge_descriptor.from_vertex][out_edge_name] = DirectJoinDescriptor(
-            from_column, to_column
-        )
-        join_descriptors[direct_edge_descriptor.to_vertex][in_edge_name] = DirectJoinDescriptor(
-            to_column, from_column
-        )
+        if isinstance(edge_descriptor, DirectEdgeDescriptor):
+            from_column = edge_descriptor.from_column
+            to_column = edge_descriptor.to_column
+            join_descriptors[edge_descriptor.from_vertex][out_edge_name] = DirectJoinDescriptor(
+                from_column, to_column
+            )
+            join_descriptors[edge_descriptor.to_vertex][in_edge_name] = DirectJoinDescriptor(
+                to_column, from_column
+            )
+        elif isinstance(edge_descriptor, CompositeEdgeDescriptor):
+            join_descriptors[edge_descriptor.from_vertex][out_edge_name] = CompositeJoinDescriptor(
+                edge_descriptor.column_pairs
+            )
+            join_descriptors[edge_descriptor.to_vertex][in_edge_name] = CompositeJoinDescriptor(
+                {
+                    (to_column, from_column)
+                    for from_column, to_column in edge_descriptor.column_pairs
+                }
+            )
+        else:
+            raise AssertionError(
+                f"Unknown edge descriptor type {edge_descriptor}: "
+                f"{type(edge_descriptor)} for edge {edge_name}."
```

Comment:
Did you mean to flip the type and the printout of the dataclass here?
```suggestion
                f"Unknown edge descriptor type {type(edge_descriptor)}: "
                f"{edge_descriptor} for edge {edge_name}."
```

Reply:
Nah, just following the `term: type` convention (see the colon)